### DISCUSSION TITLE
CI/CD ideas

### DISCUSSION THREAD

**Discussion by iMacTia** [rubyatscale/code_ownership#16] [Show and tell]
Hi all,

I bumped into this gem and it looks really great!
There are all sorts of applications for it, but I'm really curious to hear how you currently use it in your CI/CD environments.
Suggestions, examples, and ideas are all welcome: how are you currently using this gem to automate your development process?

Here is a list of interesting applications that I'm not yet sure how to implement:
1. Automatically keep the `CODEOWNERS` file up-to-date
2. Integrate the `for_stacktrace` with error tracking tools to automatically assign owners to errors

**Comment by alexevanczuk** [comment]
Thanks for starting this discussion @iMacTia . Since accountability and ownership is such a fundamental component to effectively managing a large distributed team, there are endless ways to use this. Here is a big laundry list of some of the ways that we use it at Gusto.

## Test segmentation
When we run our front-end tests, we identify the owner of each file using `CodeOwnership.for_file` to segment test result output by team

## General code ownership validations
`bin/codeownership validate --skip-autocorrect` runs all validations that ship with code ownership today, including:
- `.github/CODEOWNERS` is up to date
- All files have owners
- No file is owned by a team that does not exist

## Packwerk integrations
We use this with packwerk to identify the team that owns a packwerk package to help direct users to the right team when packwerk fails

## Logging/monitoring/observability

We use `CodeOwnership.for_class` and `CodeOwnership.for_backtrace` to tag errors/log lines/sidekiq jobs/etc. to teams. How this hooks in depends on your system's configuration and assumptions around logging and error tracking!

## Flaky test reporting

When a test flakes, we identify the owner with code ownership and report it to the team's slack channel (as specified in ther team YML).

## Paging
We use code ownership to identify the team to page when something goes wrong.

## Metric segmentation

When we have metrics (e.g. sorbet/typescript stats/modularization stats/etc.), we use CodeOwnership to segment so teams can track progress individually.
